chore: update web vitals monitoring doc (#1796)
SigNoz/signoz.io: data/docs/frontend-monitoring/web-vitals-with-traces.mdx

</Tabs>

<details>
<summary> Read more about the dependencies </summary>
- **`web-vitals`**: Core library for measuring and reporting web vitals metrics (LCP, INP, CLS, TTFB, FCP)
- **`@opentelemetry/api`**: Provides the core OpenTelemetry API for creating traces and spans
- **`@opentelemetry/sdk-trace-web`**: Web-specific implementation of OpenTelemetry tracing for browser environments
- **`@opentelemetry/instrumentation`**: Base package for OpenTelemetry instrumentation capabilities
- **`@opentelemetry/exporter-trace-otlp-http`**: Exports traces to OpenTelemetry Protocol (OTLP) over HTTP
- **`@opentelemetry/context-zone`**: Manages context propagation using browser's Zone.js for async operations
- **`@opentelemetry/resources`**: Defines service metadata and attributes for your application
- **`ua-parser-js`**: Parses user agent strings to extract browser, device, and OS information for RUM features
</details>

### Step 3: Create an instrumentation file

The `instrument.ts` file is required to setup the `OTLPTraceExporter` and `WebTracerProvider` which are used to capture web vitals as traces within your application and export them to your collector.

<Tabs entityName="plans">

>   <TabItem value="cloud" label="SigNoz Cloud" default>

Reviewer: ```suggestion
  <TabItem value="signoz-cloud" label="SigNoz Cloud" default>
```

    import {
      WebTracerProvider,
      BatchSpanProcessor,
    } from '@opentelemetry/sdk-trace-web';
    import { ZoneContextManager } from '@opentelemetry/context-zone';
    import { resourceFromAttributes } from '@opentelemetry/resources';
    import CustomSpanProcessor from './custom_span_processor';

    const exporter = new OTLPTraceExporter({
      url: `https://ingest.<<INGESTION_REGION>>.signoz.cloud:443/v1/traces`,
      headers: {
        'signoz-ingestion-key': <<INGESTION_KEY>>,
      },
    });

    const provider = new WebTracerProvider({
      resource: resourceFromAttributes({
        'service.name': 'web-vitals-demo-app-testing-2',
      }),
      spanProcessors: [new BatchSpanProcessor(exporter), CustomSpanProcessor],
    });

    provider.register({
      contextManager: new ZoneContextManager(),
    });
    ```

    - Set the `<<INGESTION_REGION>>` to match your SigNoz Cloud [region](https://signoz.io/docs/ingestion/signoz-cloud/overview/#endpoint)
    - Replace `<<INGESTION_KEY>>` with your SigNoz [ingestion key](https://signoz.io/docs/ingestion/signoz-cloud/keys/)
    - `<<SERVICE_NAME>>` is the name of your service
  </TabItem>

>   <TabItem value="self-hosted" label="SigNoz Self-Hosted">

Reviewer: ```suggestion
  <TabItem value='self-host' label='Self-Host'>
```

    span.setAttribute(CONSTANTS.BROWSER_VERSION, browserInfo.browserVersion);
  },
  onEnd: () => Promise.resolve(),
  forceFlush: () => Promise.resolve(),
  shutdown: () => Promise.resolve(),
};

export default CustomSpanProcessor;
```

Update your `instrument.ts` file to include this processor.

```ts
const provider = new WebTracerProvider({
  resource: resourceFromAttributes({
    'service.name': 'text-generator',
  }),
  spanProcessors: [new BatchSpanProcessor(exporter), CustomSpanProcessor],
});
```

Now every span exported will include these additional contextual attributes.

<figure data-zoomable align="center">
  <img
    className="box-shadowed-image"
    src="/img/guides/frontend-monitoring/web-vitals/web-vital-trace-details.webp"
    alt="Web Vitals with custom attributes"
  />
  <figcaption>
    <i>Web Vitals with custom attributes</i>
  </figcaption>
</figure>

### Setup Dashboard and alerts
You can create a [custom dashboard](https://signoz.io/docs/userguide/manage-dashboards/#steps-to-create-a-custom-dashboard) and [set alerts](https://signoz.io/docs/setup-alerts-notification/)
to monitor your core web-vitals and be notified about anything critical.

> You can find custom dashboard templates for web vitals [here](https://github.com/SigNoz/dashboards/tree/main/frontend-monitoring)

Reviewer: ```suggestion
```

Since we are mentioning it below.

<figure data-zoomable align='center'>
<img className="box-shadowed-image" src="/img/events/launch-week-1/web-vitals.webp" alt="Core Web Vitals Monitoring in SigNoz"/>
<figcaption><i>Core Web Vitals Monitoring in SigNoz </i></figcaption>
</figure>

One of the cool feature of SigNoz charts is that you can create thresholds in the chart with different color coding. For example, in the below graph there are three thresholds for `good`, `needs improvement`, and `critical` level.

<figure data-zoomable align='center'>
<img className="box-shadowed-image" src="/img/events/launch-week-1/thresholds-chart.webp" alt="Thresholds in SigNoz"/>
<figcaption><i>Thresholds provide a visual cue about your application's performance </i></figcaption>
</figure>

> You can find and try the demo application [here](https://github.com/SigNoz/frontend-monitoring-examples/tree/main/web-vitals) and also use our [custom dashboards](https://github.com/SigNoz/frontend-monitoring-examples/tree/main/dashboard-templates).

Reviewer: ```suggestion

### Out of Box Dashboard
You can [import](https://signoz.io/docs/dashboards/import-dashboard/) our out-of-the-box [Dashboard for web vitals monitoring](https://github.com/SigNoz/frontend-monitoring-examples/tree/main/dashboard-templates)

## Demo Application

We have set up a [Demo React application](https://github.com/SigNoz/frontend-monitoring-examples/tree/main/web-vitals) that demonstrates monitoring Core Web Vitals using OpenTelemetry and sending them to SigNoz.
```

---
date: 2024-08-02
title: Web Vitals Monitoring with Traces
id: web-vitals-with-traces
---

This documentation provides a step-by-step guide to setting up web vitals monitoring with traces using SigNoz and OpenTelemetry.

## Prerequisites

- [SigNoz Cloud](https://signoz.io/teams/) or self-hosted account
- A web application from where you want to send web vitals

## Setup

### Step 1: Setup OTel Collector
Install Otel collector binary using [these instructions](https://signoz.io/docs/tutorial/opentelemetry-binary-usage-in-virtual-machine/).

You would also need to update the collector config to whitelist the frontend domain. This is required to allow Cross-Origin Resource Sharing (CORS) requests from your frontend application to the OpenTelemetry collector.

```sh
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
        cors:
          allowed_origins:
            - <<YOUR_FRONTEND_URL>>
          allowed_headers: ['*']
```

### Step 2: Install dependencies

<Tabs groupId="web-vitals-dependencies">

>   <TabItem value="npm" label="Npm">

Reviewer: ```suggestion
  <TabItem value="npm" label="Npm" default>
```

    ```sh
      npm install web-vitals
      npm install @opentelemetry/api
      npm install @opentelemetry/sdk-trace-web
      npm install @opentelemetry/instrumentation
      npm install @opentelemetry/exporter-trace-otlp-http
      npm install @opentelemetry/context-zone
      npm install @opentelemetry/resources
      npm install ua-parser-js
    ```
  </TabItem>

>   <TabItem value="yarn" label="Yarn" default>

Reviewer: ```suggestion
  <TabItem value="yarn" label="Yarn">
```

    import {
      WebTracerProvider,
      BatchSpanProcessor,
    } from '@opentelemetry/sdk-trace-web';
    import { ZoneContextManager } from '@opentelemetry/context-zone';
    import { resourceFromAttributes } from '@opentelemetry/resources';
    import CustomSpanProcessor from './custom_span_processor';

    const exporter = new OTLPTraceExporter({
      url: `<<SIGNOZ_COLLECTOR_URL>>/v1/traces`,
    });

    const provider = new WebTracerProvider({
      resource: resourceFromAttributes({
        'service.name': 'web-vitals-demo-app-testing-2',
      }),
      spanProcessors: [new BatchSpanProcessor(exporter), CustomSpanProcessor],
    });

    provider.register({
      contextManager: new ZoneContextManager(),
    });
    ```

    - `<<SERVICE_NAME>>` is the name of your service
    - `<<SIGNOZ_COLLECTOR_URL>>` is the url where your collector is running
  </TabItem>
</Tabs>

### Step 4: Capture and export Web Vitals
Use the `web-vitals` library to capture key performance metrics like CLS, FID, LCP, TTFB, and FCP. 
These are captured through callback functions that you define and exported as traces.

<Tabs entityName="plans">

>   <TabItem value="ts" label="TypeScript">

Reviewer: ```suggestion
  <TabItem value="ts" label="TypeScript" default>
```

    function createWebVitalsSpan(metric: Metric) {
      const ctx = createWebVitalsContext();
      if (!ctx) {
        console.warn('Web vitals context not available for:', metric.name);
        return;
      }

      const now = hrTime();
      const webVitalsSpan = webVitalsTracer.startSpan(
        metric.name,
        { startTime: now },
        ctx
      );

      webVitalsSpan.setAttributes({
        'web_vital.name': metric.name,
        'web_vital.id': metric.id,
        'web_vital.navigationType': metric.navigationType,
        'web_vital.delta': metric.delta,
        'web_vital.rating': metric.rating,
        'web_vital.value': metric.value,
        'url.full': window.location.href,
        'page.title': document.title,
      });

      webVitalsSpan.end();
      return webVitalsSpan;
    }

    createWebVitalsContext();

    onFCP(metric => {
      createWebVitalsSpan(metric);
    });

    onINP(metric => {
      createWebVitalsSpan(metric);
    });

    onTTFB(metric => {
      createWebVitalsSpan(metric);
    });

    onLCP(metric => {
      createWebVitalsSpan(metric);
    });

    onCLS(metric => {
      createWebVitalsSpan(metric);
    });
    ```
  </TabItem>

>   <TabItem value="js" label="JavaScript" default>

Reviewer: ```suggestion
  <TabItem value="js" label="JavaScript">
```

    });

    onINP(metric => {
      createWebVitalsSpan(metric);
    });

    onTTFB(metric => {
      createWebVitalsSpan(metric);
    });

    onLCP(metric => {
      createWebVitalsSpan(metric);
    });

    onCLS(metric => {
      createWebVitalsSpan(metric);
    });
    ```
  </TabItem>
</Tabs>

### Step 6: Viewing captured Web Vitals

The captured web vitals can be viewed in the [Traces Explorer](https://signoz.io/docs/product-features/trace-explorer/). Each `web-vital` trace contains the 5 web vitals as child spans.

<figure data-zoomable align="center">
  <img
    className="box-shadowed-image"
    src="/img/guides/frontend-monitoring/web-vitals/web-vitals-traces-explorer.webp"
    alt="Web Vitals"
  />
  <figcaption>
    <i>Web Vitals</i>
  </figcaption>
</figure>

### Attaching additional identifiers to your Web Vitals

You can enrich web vitals with additional metadata like browser type, user ID, page URL etc. to enable real user monitoring (RUM)-like insights.

To do so, you need to write a custom implementation of `SpanProcessor` which will intercept all your exported spans and attach additional attributes to them.

<Tabs entityName="plans">

>   <TabItem value="ts" label="TypeScript">

Reviewer: ```suggestion
  <TabItem value="ts" label="TypeScript" default>
```

      return {
        userAgentOriginal: result.ua || '',
        userAgentVersion: result.browser.version || '',
        userAgentName: result.browser.name || '',
      };
    }

    function getUserId() {
      // You can add your custom user ID tracking logic here as well.
      // This example uses localStorage.
      const userId = localStorage.getItem('userId');
      return {
        userId: userId || '',
      };
    }

    const CustomSpanProcessor: SpanProcessor = {
      onStart: span => {
        const userData = getUserId();
        const browserInfo = getBrowserInfo();

        span.setAttribute(CONSTANTS.USER_ID, userData.userId);
        span.setAttribute(
          CONSTANTS.USER_AGENT_ORIGINAL,
          browserInfo.userAgentOriginal
        );
        span.setAttribute(
          CONSTANTS.USER_AGENT_VERSION,
          browserInfo.userAgentVersion
        );
        span.setAttribute(CONSTANTS.USER_AGENT_NAME, browserInfo.userAgentName);
      },
      onEnd: () => Promise.resolve(),
      forceFlush: () => Promise.resolve(),
      shutdown: () => Promise.resolve(),
    };

    export default CustomSpanProcessor;
    ```
  </TabItem>

>   <TabItem value="js" label="JavaScript" default>

Reviewer: ```suggestion
  <TabItem value="js" label="JavaScript">
```

    ```js:instrumentation.js
    import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
    import {
      WebTracerProvider,
      BatchSpanProcessor,
    } from '@opentelemetry/sdk-trace-web';
    import { ZoneContextManager } from '@opentelemetry/context-zone';
    import { resourceFromAttributes } from '@opentelemetry/resources';

    const exporter = new OTLPTraceExporter({
      // For the self-hosted version, please use the collector url instead.
      url: `https://ingest.<INGESTION_REGION>.signoz.cloud:443/v1/traces`,
      headers: {
        'signoz-ingestion-key': '<INGESTION_KEY>',
      },
    });

    const provider = new WebTracerProvider({
      resource: resourceFromAttributes({
        'service.name': '<SERVICE_NAME>',
      }),
      spanProcessors: [new BatchSpanProcessor(exporter)],
    });

    provider.register({
      contextManager: new ZoneContextManager(),
    });
    ```

    - Set the `<INGESTION_REGION>` to match your SigNoz Cloud [region](https://signoz.io/docs/ingestion/signoz-cloud/overview/#endpoint)
    - Replace `<INGESTION_KEY>` with your SigNoz [ingestion key](https://signoz.io/docs/ingestion/signoz-cloud/keys/)

>     - `<SERVICE_NAME>` is the name of your service

Author: Not required

    onFCP(metric => {
      createWebVitalsSpan(metric);
    });

    onINP(metric => {
      createWebVitalsSpan(metric);
    });

    onTTFB(metric => {
      createWebVitalsSpan(metric);
    });

    onLCP(metric => {
      createWebVitalsSpan(metric);
    });

    onCLS(metric => {
      createWebVitalsSpan(metric);
    });
    ```

  </TabItem>
</Tabs>

## Validate

The captured web vitals can be viewed in the [Traces Explorer](https://signoz.io/docs/product-features/trace-explorer/). Each `web-vital` trace contains the 5 web vitals as child spans.

<figure data-zoomable align="center">
  <img
    className="box-shadowed-image"
    src="/img/guides/frontend-monitoring/web-vitals/web-vitals-traces-explorer.webp"
    alt="Web Vitals"
  />
  <figcaption>
    <i>Web Vitals</i>
  </figcaption>
</figure>

### Attaching additional identifiers to your Web Vitals

You can enrich web vitals with additional metadata like browser type, user ID, page URL etc. to enable real user monitoring (RUM)-like insights.

To do so, you need to write a custom implementation of `SpanProcessor` which will intercept all your exported spans and attach additional attributes to them.

> <Tabs entityName="plans">

Reviewer 2: Nit: plans entity name is not valid here, it should be language

  <TabItem value="ts" label="TypeScript" default>

>     ```ts

Reviewer 2: File name missing similar to above

>     ```ts

Reviewer 2: File name missing similar to above
